drivers/usb/udc/udc_it82xx2.c

```diff
@@ -0,0 +1,1434 @@
+
+	usb_regs->dc_address = DC_ADDR_NULL;
+
+	return 0;
+}
+
+static void it82xx2_enable_sof_int(const struct device *dev, bool enable)
+{
+	const struct usb_it82xx2_config *config = dev->config;
+	struct usb_it82xx2_regs *const usb_regs = config->base;
+
+	usb_regs->dc_interrupt_status = DC_SOF_RECEIVED;
+	if (enable) {
+		usb_regs->dc_interrupt_mask |= DC_SOF_RECEIVED;
+	} else {
+		usb_regs->dc_interrupt_mask &= ~DC_SOF_RECEIVED;
+	}
+}
+
+void it82xx2_dc_reset(const struct device *dev)
+{
+	const struct usb_it82xx2_config *config = dev->config;
+	struct usb_it82xx2_regs *const usb_regs = config->base;
+	struct it82xx2_usb_ep_regs *ep_regs = usb_regs->usb_ep_regs;
+	struct it82xx2_usb_ep_fifo_regs *ff_regs = usb_regs->fifo_regs;
+	struct it82xx2_priv_data *priv = udc_get_private(dev);
+
+	for (uint8_t ep_idx = 0; ep_idx < 4; ep_idx++) {
+		ff_regs[ep_idx].ep_rx_fifo_ctrl = FIFO_FORCE_EMPTY;
+		ff_regs[ep_idx].ep_tx_fifo_ctrl = FIFO_FORCE_EMPTY;
+	}
+
+	ep_regs[0].ep_ctrl.value = ENDPOINT_EN;
+	usb_regs->dc_address = DC_ADDR_NULL;
+	usb_regs->dc_interrupt_status = DC_NAK_SENT_INT | DC_SOF_RECEIVED;
+
+	priv->fifo_ready[0] = false;
+	priv->fifo_ready[1] = false;
+	priv->fifo_ready[2] = false;
+	k_sem_give(&priv->fifo_sem[0]);
+	k_sem_give(&priv->fifo_sem[1]);
+	k_sem_give(&priv->fifo_sem[2]);
+}
+
+static int it82xx2_xfer_in_data(const struct device *dev, uint8_t ep, struct net_buf *buf)
+{
+	const uint8_t ep_idx = USB_EP_GET_IDX(ep);
+	const struct usb_it82xx2_config *config = dev->config;
+	struct usb_it82xx2_regs *const usb_regs = config->base;
+	struct it82xx2_usb_ep_fifo_regs *ff_regs = usb_regs->fifo_regs;
+	struct it82xx2_priv_data *priv = udc_get_private(dev);
+	struct udc_ep_config *ep_cfg = udc_get_ep_cfg(dev, ep);
+	uint8_t ep_fifo = (ep_idx > 0) ? (ep_fifo_res[ep_idx % SHARED_FIFO_NUM]) : 0;
+	size_t len;
+
+	if (ep_idx == 0) {
+		ff_regs[ep_idx].ep_tx_fifo_ctrl = FIFO_FORCE_EMPTY;
+	} else {
+		k_sem_take(&priv->fifo_sem[ep_fifo - 1], K_FOREVER);
+		it82xx2_usb_fifo_ctrl(dev, ep);
+	}
+
+	len = MIN(buf->len, ep_cfg->mps);
+
+	for (size_t i = 0; i < len; i++) {
+		ff_regs[ep_fifo].ep_tx_fifo_data = buf->data[i];
+	}
+
+	if (IT8XXX2_IS_EXTEND_ENDPOINT(ep_idx)) {
+		it82xx2_usb_extend_ep_ctrl(dev, ep_idx, EP_READY_ENABLE, true);
+	}
+	it82xx2_usb_set_ep_ctrl(dev, ep_fifo, EP_READY_ENABLE, true);
+
+	if (ep_idx != 0) {
+		priv->fifo_ready[ep_fifo - 1] = true;
+	}
+
+	LOG_DBG("Writed %d packets to endpoint%d tx fifo", buf->len, ep_idx);
+
+	return 0;
+}
+
+static int it82xx2_xfer_out_data(const struct device *dev, uint8_t ep, struct net_buf *buf)
+{
+	const struct usb_it82xx2_config *config = dev->config;
+	struct usb_it82xx2_regs *const usb_regs = config->base;
+	struct it82xx2_usb_ep_regs *ep_regs = usb_regs->usb_ep_regs;
+	struct it82xx2_usb_ep_fifo_regs *ff_regs = usb_regs->fifo_regs;
+	const uint8_t ep_idx = USB_EP_GET_IDX(ep);
+	uint8_t ep_fifo = (ep_idx > 0) ? (ep_fifo_res[ep_idx % SHARED_FIFO_NUM]) : 0;
```

Comment:
I do like this type of obfuscated logic at the top of a block. IMO better is to write it below variable definition block. btw, overparenthesesing, ep_fifo should probably be fifo_idx because it is not endpoint FIFO.

```c
uint8_t fifo_idx;

fifo_idx = ep_idx > 0 ? ep_fifo_res[ep_idx % SHARED_FIFO_NUM] : 0;
```

or maybe
 
```c
const uint8_t fifo_idx = get_fifo_idx(dev, ep);
```

Reply:
done.